### PR TITLE
fix(poker): attribute gpu rewards to acting seat

### LINKED ISSUE #12: PokerGPU rewards can depend on the next player's equity instead of the acting player
## Intent
`PokerGPU.step()` advances `self.idx` to the next player before calling `poker_reward_gpu()`. The reward function then reads `self.equities[self.g, self.idx]`, so the returned reward can depend on the next seat's hand instead of the actor who actually took the action.

## Problem Description
Current behavior:
- `step()` snapshots stack/bet state for the acting seat, executes the action, advances `self.idx`, and only then computes rewards.
- `poker_reward_gpu()` indexes equity and status using the current `self.idx`, which may already point to the next player.
- As a result, the same action by the same acting player can receive different rewards if only the next player's hole cards change.

Expected behavior:
- Reward for an action should be computed from the acting player's pre/post-action state, not from whichever seat is next to act.

Conditions:
- This manifests on any non-terminal step where `self.idx` changes before reward calculation.
- It is especially visible postflop, where hand equities differ materially across seats.

## Affected Files
- `environments/Poker/PokerGPU.py` - advances `self.idx` before reward calculation and uses `self.idx` inside the reward function.
- `scripts/Poker/trainGPU.py` - feeds the resulting rewards directly into Q-learning updates.

## Code Snippets
```python
# environments/Poker/PokerGPU.py - lines 452-458
prev_done = self.is_done.clone()
self.prev_stacks.copy_(self.stacks[self.g, self.idx])
self.prev_invested.copy_(self.current_round_bet[self.g, self.idx])

# 1) calculate the equties of players hands
self.equities.fill_(.5)
self.calculate_equities()
```
These buffers are captured for the acting seat before the action is applied.

```python
# environments/Poker/PokerGPU.py - lines 485-487
continue_mask = ~self.is_round_over & has_next_player
if continue_mask.any():
    self.idx[continue_mask] = candidate_seats[continue_mask, next_positions[continue_mask]]
```
The acting seat can be replaced with the next player before reward calculation.

```python
# environments/Poker/PokerGPU.py - lines 540-543
rewards = self.poker_reward_gpu(actions=actions)
rewards[prev_done] = 0
return self.get_obs(), rewards, self.is_done, self.is_truncated, self.get_info()
```
Reward is computed after `self.idx` may have been mutated.

```python
# environments/Poker/PokerGPU.py - lines 268-287
active_counts = ((self.status == self.ACTIVE) | (self.status == self.ALLIN)).sum(dim=1).float()
fair_shares = 1.0 / torch.clamp(active_counts, min=1.0)
call_costs = torch.maximum(torch.zeros_like(self.highest), self.highest - self.prev_invested)

# first part of reward function, promote higher equities with bigger pots
e = self.equities[self.g, self.idx]
...
self.s[call_mask] = (e[call_mask] - o[call_mask]) * self.pots[call_mask]
```
The reward function reads equity from the possibly updated `self.idx`, even though `self.prev_invested` still refers to the actor who just moved.

```python
# scripts/Poker/trainGPU.py - lines 54-60
agents[agent_types.index(PokerAgentType.QLEARNING)].train_step(
    states=state[active_games],
    actions=actions[active_games],
    rewards=rewards[active_games],
    next_states=next_state[active_games],
    dones=dones[active_games]
)
```
The corrupted reward is consumed directly by the learning update.

## Root Cause Analysis
The root cause is ordering inside `environments/Poker/PokerGPU.py`. `step()` records actor-specific buffers, but then mutates `self.idx` during next-player selection before calling `poker_reward_gpu()`. Inside `poker_reward_gpu()`, the reward formula reads `self.equities[self.g, self.idx]`, which now points at the next seat for continuing hands. That mixes actor-local quantities (`prev_invested`, `actions`) with the wrong player's equity and status.

## Impact Assessment
- Affected flows: essentially every non-terminal postflop action in the GPU environment.
- User impact: rewards can flip sign or magnitude based on another player's cards.
- Data integrity risk: critical. This directly corrupts the TD targets used for Q-learning.
- Failure mode: silent. Training proceeds with numerically plausible but semantically wrong rewards.
- Severity: Critical. Reward attribution is a core RL invariant, and this bug violates it on the main action path.

## Potential Fixes
**Option A - Capture the acting seat and pass it into `poker_reward_gpu()`**
Store `actor_idx = self.idx.clone()` before any mutation, then compute `e = self.equities[self.g, actor_idx]`. This is the minimal targeted fix.

**Option B - Compute reward before mutating turn order**
Move reward calculation earlier in `step()`, before `self.idx` is updated. This keeps actor context intact, but requires care around terminal payouts and post-action pot values.

**Option C - Split reward inputs into an explicit actor-state struct**
Pass the actor seat, actor equity, and actor pre/post-action quantities as explicit tensors into the reward function. This is a larger refactor but eliminates hidden dependence on mutable environment cursors.

## Reproduction Steps
1. Instantiate `PokerGPU` with `n_players=3`, `max_players=3`, `n_games=1`.
2. Force a flop state where seat `0` is acting and seat `1` will act next:
   - `stages = [1]`
   - board flop = `Ah Kh 2c`
   - `status = [ACTIVE, ACTIVE, FOLDED]`
   - `idx = 0`
   - `agg = 1`
   - `highest = 5`
   - `pots = 10`
3. Keep seat `0`'s hand fixed as `Qh Jh` and call `env.step(torch.tensor([1]))`.
4. Repeat from the same state, changing only seat `1`'s hand:
   - weak next hand: `7d 8s`
   - strong next hand: `Th 9h`
5. Actual result:
   - weak next hand reward: about `20.34`
   - strong next hand reward: about `-28.04`
6. Expected result: the reward for seat `0`'s call should not change when only seat `1`'s hole cards change.

## Additional Context
- Preflop actions are harder to diagnose because preflop equity is currently hard-coded to `0.5`, which masks the attribution bug.
- This bug is independent of the previously reported post-transition actor issue; it appears even on ordinary intra-street turn advancement.

### PR DESCRIPTION
**Pull request by cerredz**
## Summary
Fix PokerGPU reward attribution so post-action rewards always use the acting seat's equity, even after `self.idx` advances to the next player or resets on a street transition.

## What Changed
- Snapshotted the acting seat in `PokerGPU.step()` before any turn-order mutation and passed it into `poker_reward_gpu()`.
- Updated `poker_reward_gpu()` to read equities from the captured actor index instead of the mutable cursor.
- Added five regression tests covering continuing actions, skipped seats, street transitions, and batched reward attribution.

## Files of Interest
- `environments/Poker/PokerGPU.py` - fixes actor-index reward attribution in the GPU environment.
- `tests/poker/test_poker_gpu_actor_reward_attribution.py` - adds targeted regression coverage for issue #12.
- `.claude/reasoning/poker_gpu_actor_reward_attribution_specification.md` - records the spec, scope, risks, and validation plan.

## How To Test
1. Run `pytest tests/poker/test_poker_gpu_actor_reward_attribution.py -q`.
2. Confirm all five tests pass and reward values stay stable when only the next player's hole cards change.
3. Run `pytest tests/poker -q` to verify the broader poker regression suite.

## Risks
This is a narrow environment-side reward fix, but it changes the reward tensor consumed by Q-learning updates. Training behavior should improve semantically, though historical reward curves may shift because corrupted targets are no longer produced.

## Linked Issues
Closes #12

### CODE REVIEW

**Comment by cerredz** [comment]
Closing as stale because the acting-seat reward attribution fix landed through later merged work.
